lib/data/repository/default_marvell_repository.dart

```diff
@@ -0,0 +1,113 @@
+import 'package:dio/dio.dart';
+import 'package:marvelapp_flutter/domain/entities/character.dart';
+import 'package:marvelapp_flutter/domain/error_handling/exceptions.dart';
+import 'package:marvelapp_flutter/domain/repositories/marvell_repository.dart';
+import 'package:marvelapp_flutter/domain/entities/series.dart';
+import 'package:marvelapp_flutter/data/data_sources/local/character_data_source.dart';
+import 'package:marvelapp_flutter/data/data_sources/remote/remote_data_source.dart';
+
+class DefaultMarvellRepository extends MarvellRepository {
+  late CharacterDataSource localDataSource;
+  late RemoteDataSource remoteDataSource;
+
+  static final DefaultMarvellRepository _repository = DefaultMarvellRepository._internal();
+
+  DefaultMarvellRepository._internal();
+
+  factory DefaultMarvellRepository(CharacterDataSource localDataSource, RemoteDataSource remoteDataSource) {
+    _repository.localDataSource = localDataSource;
+    _repository.remoteDataSource = remoteDataSource;
+    return _repository;
+  }
+
+  @override
+  Future<Character> getCharacterDetail(String characterId) async {
+    try {
+      var httpResponse = await remoteDataSource.getCharacterDetail(characterId);
+      if (httpResponse.response.statusCode == 200) {
+        var data = httpResponse.data.data;
+        Character? tempCharacter = data?.results?.map((item) => item.fromApiToCharacter("standard_xlarge")).single;
+        Character character = tempCharacter as Character;
+        return character;
+      } else {
+        throw DataRetrieveException();
+      }
+    } on DioError catch (error) {
+      throw _getException(error);
+    }
+  }
+
+  @override
+  Future<List<Character>> getCharacters([int offset = 0]) async {
+    if (offset == 0) {
+      List<Character> characters = List.empty();
+      try {
+        characters = await _getRemote(offset);
+      } on Exception {
+        characters = await _getLocal();
+      }
+      return characters;
+    }
+    try {
+      return _getRemote(offset);
+    } on DioError catch (error) {
+      throw _getException(error);
+    }
+  }
+
+  Future<List<Character>> _getRemote(int offset) async {
+    List<Character> characters = List.empty();
+    try {
+      var httpResponse = await remoteDataSource.getCharacters(offset);
+      if (httpResponse.response.statusCode == 200) {
+        var data = httpResponse.data.data;
+        List<Character>? tempCharacters =
+            data?.results?.map((item) => item.fromApiToCharacter("standard_medium")).toList();
+        characters = tempCharacters ?? characters;
+        await _saveCharactersInDatabase(characters);
```

Comment:
Should not we store only the first page?

Reply:
add condition. done